onnxruntime/python/tools/symbolic_shape_infer.py

```diff
@@ -2080,29 +2108,67 @@ def _infer_MultiHeadAttention(self, node):
+        total_sequence_length = None
+        output_dtype = None
         if query_shape is not None:
             if len(query_shape) == 3:
-                key_shape = self._get_shape(node, 1)
+                key_shape = self._try_get_shape(node, 1)
                 # By default, hidden size is same for Q/K/V. Only need check v_hidden_size when value is provided.
                 output_shape = query_shape
-                if key_shape and len(key_shape) == 3:
-                    value_shape = self._get_shape(node, 2)
-                    if value_shape and len(value_shape) == 3:
+                if key_shape is not None and len(key_shape) == 3:
+                    value_shape = self._try_get_shape(node, 2)
+                    if value_shape is not None and len(value_shape) == 3:
                         output_shape[2] = value_shape[2]
+                    total_sequence_length = key_shape[1]
 
                 output_dtype = self.known_vi_[node.input[0]].type.tensor_type.elem_type
                 vi = self.known_vi_[node.output[0]]
                 vi.CopyFrom(helper.make_tensor_value_info(node.output[0], output_dtype, output_shape))
+
             elif len(query_shape) == 5:
                 if isinstance(query_shape[2], int) and isinstance(query_shape[4], int):
                     output_shape = [query_shape[0], query_shape[1], query_shape[2] * query_shape[4]]
                 else:
                     output_shape = [query_shape[0], query_shape[1], f"{query_shape[2]}*{query_shape[4]}"]
 
+                total_sequence_length = query_shape[1]
+
                 output_dtype = self.known_vi_[node.input[0]].type.tensor_type.elem_type
                 vi = self.known_vi_[node.output[0]]
                 vi.CopyFrom(helper.make_tensor_value_info(node.output[0], output_dtype, output_shape))
 
+            if len(node.output) > 1:
+                batch_size = query_shape[0]
+                num_heads = get_attribute(node, "num_heads")
+
+                head_size = None
+                if len(query_shape) == 3:
+                    head_size = (
+                        int(query_shape[2] / num_heads)
+                        if isinstance(query_shape[2], int)
+                        else f"{query_shape[2]}/{num_heads}"
+                    )
+                else:
+                    head_size = query_shape[4]
+
+                past_shape = self._try_get_shape(node, 5)
+
+                if past_shape is not None:
+                    if isinstance(past_shape[2], int) and isinstance(total_sequence_length, int):
+                        total_sequence_length = past_shape[2] + total_sequence_length
+                    else:
+                        total_sequence_length = f"{past_shape[2]}+{total_sequence_length}"
+
+                # Keep the batch size and total sequence length as string
+                present_shape = [str(batch_size), num_heads, str(total_sequence_length), head_size]
```

Comment:
I think we can remove str() here. Did you encounter issue without explicitly converting to string?

Reply:
removed. actually found a bug in retrieving past_key from mha. the index was fixed.